mixer/pkg/checkcache/cache.go

```diff
@@ -0,0 +1,198 @@
+// cache then its capacity will cause eviction of older entries.
+func New(capacity int32) *CheckCache {
+	cc := &CheckCache{
+		cache:       cache.NewLRU(time.Minute*60, 1*time.Minute, capacity),
+		globalWords: attribute.GlobalList(),
+	}
+
+	cc.cacheWrites = prometheus.NewCounterFunc(prometheus.CounterOpts{
+		Namespace: "mixer",
+		Subsystem: "checkcache",
+		Name:      "cache_writes",
+		Help:      "The number of times state in the cache was added or updated.",
+	}, func() float64 {
+		s := cc.cache.Stats()
+		return float64(s.Writes)
+	})
+
+	cc.cacheHits = prometheus.NewCounterFunc(prometheus.CounterOpts{
+		Namespace: "mixer",
+		Subsystem: "checkcache",
+		Name:      "cache_hits",
+		Help:      "The number of times a cache lookup operation succeeded to find an entry in the cache.",
+	}, func() float64 {
+		s := cc.cache.Stats()
+		return float64(s.Hits)
+	})
+
+	cc.cacheMisses = prometheus.NewCounterFunc(prometheus.CounterOpts{
+		Namespace: "mixer",
+		Subsystem: "checkcache",
+		Name:      "cache_misses",
+		Help:      "The number of times a cache lookup operation failed to find an entry in the cache.",
+	}, func() float64 {
+		s := cc.cache.Stats()
+		return float64(s.Misses)
+	})
+
+	cc.cacheEvictions = prometheus.NewCounterFunc(prometheus.CounterOpts{
+		Namespace: "mixer",
+		Subsystem: "checkcache",
+		Name:      "cache_evictions",
+		Help:      "The number of entries that have been evicted from the cache.",
+	}, func() float64 {
+		s := cc.cache.Stats()
+		return float64(s.Evictions)
+	})
+
+	_ = prometheus.Register(cc.cacheWrites)
+	_ = prometheus.Register(cc.cacheHits)
+	_ = prometheus.Register(cc.cacheMisses)
+	_ = prometheus.Register(cc.cacheEvictions)
+
+	return cc
+}
+
+// Close releases any resources used by the check cache.
+func (cc *CheckCache) Close() error {
+	_ = prometheus.Unregister(cc.cacheWrites)
+	_ = prometheus.Unregister(cc.cacheHits)
+	_ = prometheus.Unregister(cc.cacheMisses)
+	_ = prometheus.Unregister(cc.cacheEvictions)
+	return nil
+}
+
+// Get looks up an attribute bag in the cache.
+func (cc *CheckCache) Get(attrs attribute.Bag) (CacheValue, bool) {
+	cc.keyShapesLock.RLock()
+	shapes := cc.keyShapes
+	cc.keyShapesLock.RUnlock()
+
+	// find a matching key shape
+	for _, shape := range shapes {
+		if shape.isCompatible(attrs) {
+
+			// given the compatible key shape, make a key
+			key := shape.makeKey(attrs)
+
+			// see if we have an entry in the cache for this key
+			if result, ok := cc.cache.Get(key); ok {
+				v := result.(CacheValue)
+				if v.Expiration.Before(time.Now()) {
+					// Entry expired. This happens because the underlying ExpiringCache only lazily cleans up
+					// expired entries. Since we want to be more precise, we do our own freshness check and throw
+					// out stale results
+					return CacheValue{}, false
+				}
+
+				// got a match!
+				return result.(CacheValue), true
```

Comment:
Also need to deduct the valid_use_count?

Reply:
ValidUseCount is a bogus concept given this caching in Mixer. It just doesn't make any sense. I'm expecting to remove this whole concept before 1.0.